File: internal/protocol/crypto.go
// Package protocol is the relay wire + crypto contract, shared by the runner and
// (mirrored in JS) the browser viewer. See PROTOCOL.md — it is the source of truth.
//
// All terminal IO is end-to-end encrypted with AES-256-GCM under keys derived from
// a 32-byte session secret S that the relay never sees. Every primitive here is Go
// standard library, so the runner needs no third-party crypto.
package protocol

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hkdf"
	"crypto/pbkdf2"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base32"
	"encoding/binary"
	"errors"
	"fmt"
)

const (
	// SecretLen is the size of the session secret S.
	SecretLen = 32
	// KeyLen is the size of each derived AES-256 key.
	KeyLen = 32

	nonceLen   = 12
	seqLen     = 8
	tagLen     = 16
	pbkdf2Iter = 600_000

	infoR2V = "relay/v1 runner->viewer"
	infoV2R = "relay/v1 viewer->runner"
	infoFP  = "relay/v1 fingerprint"
)

// Errors returned when opening a frame. They are deliberately coarse: a peer (or
// the relay) must not learn why a frame failed to open.
var (
	ErrFrameShort = errors.New("protocol: frame too short")
	ErrOpen       = errors.New("protocol: open failed")
)

// Keys are the directional AEAD keys plus a short fingerprint, all derived from the
// session secret (and an optional passphrase). The fingerprint is shown at both ends
// so a human can confirm they are looking at the same session.
type Keys struct {
	R2V         []byte // seals runner→viewer
	V2R         []byte // seals viewer→runner
	Fingerprint string
}

// DeriveKeys turns the session secret (and optional passphrase) into directional
// keys. id is the session id, used as the HKDF salt so keys are bound to the session.
// When a passphrase is set it is stretched with PBKDF2 and folded into the key
// material, so knowing the link alone (the secret) is not enough to decrypt.
func DeriveKeys(secret []byte, id, passphrase string) (Keys, error) {
	if len(secret) != SecretLen {
		return Keys{}, fmt.Errorf("protocol: secret must be %d bytes, got %d", SecretLen, len(secret))
	}
	ikm := secret
	if passphrase != "" {
		pw, err := pbkdf2.Key(sha256.New, passphrase, []byte(id), pbkdf2Iter, KeyLen)
		if err != nil {
			return Keys{}, err
		}
		ikm = append(append(make([]byte, 0, len(secret)+len(pw)), secret...), pw...)
	}
	salt := []byte(id)
	r2v, err := hkdf.Key(sha256.New, ikm, salt, infoR2V, KeyLen)
	if err != nil {
		return Keys{}, err
	}
	v2r, err := hkdf.Key(sha256.New, ikm, salt, infoV2R, KeyLen)
	if err != nil {
		return Keys{}, err
	}
	fp, err := hkdf.Key(sha256.New, ikm, salt, infoFP, 10)
	if err != nil {
		return Keys{}, err
	}
	return Keys{
		R2V:         r2v,
		V2R:         v2r,
		Fingerprint: base32.StdEncoding.WithPadding(base32.NoPadding).EncodeToString(fp),
	}, nil
}

// Cipher seals and opens frames for one direction (one key). aad binds frames to the
// session (it is the session id), preventing cross-session replay.
type Cipher struct {
	aead cipher.AEAD
	aad  []byte
}

// NewCipher builds a Cipher from a 32-byte key and the additional authenticated data.
func NewCipher(key, aad []byte) (*Cipher, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	return &Cipher{aead: aead, aad: aad}, nil
}

// Seal produces a wire frame (nonce ‖ ciphertext) for the message (seq, kind,
// payload). The nonce is fresh random bytes, so reconnects can never reuse a nonce.
func (c *Cipher) Seal(seq uint64, kind byte, payload []byte) ([]byte, error) {
	nonce := make([]byte, nonceLen)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	return c.sealWithNonce(nonce, seq, kind, payload), nil
}

// sealWithNonce is Seal with a caller-supplied nonce — used only by golden vectors.
func (c *Cipher) sealWithNonce(nonce []byte, seq uint64, kind byte, payload []byte) []byte {
	pt := make([]byte, seqLen+1+len(payload))
	binary.BigEndian.PutUint64(pt[:seqLen], seq)
	pt[seqLen] = kind
	copy(pt[seqLen+1:], payload)
	frame := make([]byte, nonceLen, nonceLen+len(pt)+tagLen)
	copy(frame, nonce)
	return c.aead.Seal(frame, nonce, pt, c.aad)
}

// Open authenticates and decrypts a wire frame, returning (seq, kind, payload). A
// frame that fails authentication (tampered, wrong key, wrong session) returns ErrOpen.
func (c *Cipher) Open(frame []byte) (seq uint64, kind byte, payload []byte, err error) {
	if len(frame) < nonceLen+seqLen+1+tagLen {
		return 0, 0, nil, ErrFrameShort
	}
	nonce := frame[:nonceLen]
	pt, err := c.aead.Open(nil, nonce, frame[nonceLen:], c.aad)
	if err != nil {
		return 0, 0, nil, ErrOpen
	}
	if len(pt) < seqLen+1 {
		return 0, 0, nil, ErrFrameShort
	}
	seq = binary.BigEndian.Uint64(pt[:seqLen])
	kind = pt[seqLen]
	payload = pt[seqLen+1:]
	return seq, kind, payload, nil
}

File: portal/priv/static/assets/crypto.js
// relay end-to-end crypto — the browser mirror of internal/protocol/crypto.go.
// Pure Web Crypto + standard globals, so it runs unchanged in a browser and under
// `node --test`. Interop with Go is pinned by internal/protocol/testdata/vectors.json.

const enc = new TextEncoder();
const subtle = globalThis.crypto.subtle;

const NONCE_LEN = 12;
const TAG_LEN = 16;
const SEQ_LEN = 8;
const PBKDF2_ITER = 600000;

const INFO_R2V = "relay/v1 runner->viewer";
const INFO_V2R = "relay/v1 viewer->runner";
const INFO_FP = "relay/v1 fingerprint";

// --- encodings ---------------------------------------------------------------

export function bytesToHex(b) {
  let s = "";
  for (const x of b) s += x.toString(16).padStart(2, "0");
  return s;
}

export function hexToBytes(h) {
  const a = new Uint8Array(h.length / 2);
  for (let i = 0; i < a.length; i++) a[i] = parseInt(h.slice(i * 2, i * 2 + 2), 16);
  return a;
}

export function b64urlToBytes(s) {
  s = s.replace(/-/g, "+").replace(/_/g, "/");
  while (s.length % 4) s += "=";
  const bin = atob(s);
  const a = new Uint8Array(bin.length);
  for (let i = 0; i < a.length; i++) a[i] = bin.charCodeAt(i);
  return a;
}

// RFC 4648 base32, no padding — matches Go base32.StdEncoding.WithPadding(NoPadding).
const B32 = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
function base32nopad(bytes) {
  let bits = 0, value = 0, out = "";
  for (const b of bytes) {
    value = (value << 8) | b;
    bits += 8;
    while (bits >= 5) {
      out += B32[(value >>> (bits - 5)) & 31];
      bits -= 5;
    }
  }
  if (bits > 0) out += B32[(value << (5 - bits)) & 31];
  return out;
}

function concat(a, b) {
  const o = new Uint8Array(a.length + b.length);
  o.set(a, 0);
  o.set(b, a.length);
  return o;
}

// --- key derivation ----------------------------------------------------------

async function hkdf(ikm, salt, info, lenBytes) {
  const k = await subtle.importKey("raw", ikm, "HKDF", false, ["deriveBits"]);
  const bits = await subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt, info: enc.encode(info) },
    k,
    lenBytes * 8,
  );
  return new Uint8Array(bits);
}

async function pbkdf2(passphrase, salt, iter, lenBytes) {
  const k = await subtle.importKey("raw", enc.encode(passphrase), "PBKDF2", false, ["deriveBits"]);
  const bits = await subtle.deriveBits(
    { name: "PBKDF2", hash: "SHA-256", salt, iterations: iter },
    k,
    lenBytes * 8,
  );
  return new Uint8Array(bits);
}

// deriveKeys mirrors protocol.DeriveKeys: directional key material + fingerprint.
export async function deriveKeys(secret, id, passphrase) {
  const salt = enc.encode(id);
  let ikm = secret;
  if (passphrase) ikm = concat(secret, await pbkdf2(passphrase, salt, PBKDF2_ITER, 32));
  const r2v = await hkdf(ikm, salt, INFO_R2V, 32);
  const v2r = await hkdf(ikm, salt, INFO_V2R, 32);
  const fp = await hkdf(ikm, salt, INFO_FP, 10);
  return { r2v, v2r, fingerprint: base32nopad(fp) };
}

// --- AEAD --------------------------------------------------------------------

// newCipher returns { seal, open } for one direction. aad is the session id bytes.
export async function newCipher(keyBytes, aad) {
  const key = await subtle.importKey("raw", keyBytes, "AES-GCM", false, ["encrypt", "decrypt"]);

  async function sealWithNonce(nonce, seq, kind, payload) {
    const pt = new Uint8Array(SEQ_LEN + 1 + payload.length);
    // seq is a JS Number; BigInt(seq) is exact up to 2^53 (open() converts back the
    // same way). Unreachable in a terminal session — see PROTOCOL.md.
    new DataView(pt.buffer).setBigUint64(0, BigInt(seq));
    pt[SEQ_LEN] = kind;
    pt.set(payload, SEQ_LEN + 1);
    const ct = new Uint8Array(
      await subtle.encrypt({ name: "AES-GCM", iv: nonce, additionalData: aad, tagLength: 128 }, key, pt),
    );
    return concat(nonce, ct);
  }

  return {
    async seal(seq, kind, payload) {
      const nonce = globalThis.crypto.getRandomValues(new Uint8Array(NONCE_LEN));
      return sealWithNonce(nonce, seq, kind, payload);
    },
    async open(frame) {
      if (frame.length < NONCE_LEN + SEQ_LEN + 1 + TAG_LEN) throw new Error("frame too short");
      const nonce = frame.subarray(0, NONCE_LEN);
      let pt;
      try {
        pt = new Uint8Array(
          await subtle.decrypt(
            { name: "AES-GCM", iv: nonce, additionalData: aad, tagLength: 128 },
            key,
            frame.subarray(NONCE_LEN),
          ),
        );
      } catch {
        throw new Error("open failed");
      }
      if (pt.length < SEQ_LEN + 1) throw new Error("frame too short");
      const dv = new DataView(pt.buffer, pt.byteOffset, pt.byteLength);
      return { seq: Number(dv.getBigUint64(0)), kind: pt[SEQ_LEN], payload: pt.subarray(SEQ_LEN + 1) };
    },
    // Deterministic seal — for golden-vector tests only.
    _sealWithNonce: sealWithNonce,
  };
}
